packages/material-ui/src/Popper/Popper.test.js

```diff
@@ -169,6 +167,40 @@ describe('<Popper />', () => {
     });
   });
 
+  describe('prop: keepMounted', () => {
+    it('should keep the children in the DOM', () => {
+      const wrapper = mount(<Popper {...defaultProps} open={false} keepMounted />);
+      assert.strictEqual(wrapper.find('span').length, 1);
+    });
+
+    it('should remove from DOM when closed on update frame immediately after mount', () => {
+      const wrapper = mount(<Popper {...defaultProps} open={false} keepMounted={false} />);
+      wrapper.setProps({ open: true });
+      // Portal has not yet re-rendered, so children do not exist on the DOM
```

Comment:
I don't think that's true. `mount` and `setProps` are wrapped in `act`. I think this has to do with `react-transition-group` handles state transitions. It requires one render cycle per transition. If you set `in={true}` the `Transition` component still has the `exited` state. The state is switched to `entering` in the next render cycle.

Reply:
There is no transition component rendered in this case. We probably wait for `forceUpdate` resolution? I'm not sure.

Reply:
You are indeed correct... which is odd, because it would appear that we're having a different issue in my workplace app :/

In any case... Why does the assertion that the `span` does not exist still pass, seeing as nothing is returning `null`, and MUI is using the "exited" state to alter the child's visibility to "hidden" in `Grow`([here](https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/Grow/Grow.js#L124)). 

`react-transition-group` only returns null (and thus doesn't render children) when it has an `UNMOUNTED` state ([here](https://github.com/reactjs/react-transition-group/blob/master/src/Transition.js#L339))

Reply:
I was irritated by the `keepMounted`. Still the comment is misleading. `Portal` should be re-rendered at this point. Could you add some assertions before the `setProps` so that we're sure we know what's happening.

Honestly since this fix is about the integration with Transition component I would actually test this integration instead of mocking behavior that might be wrong.

Reply:
I've put a load of logging into the following test-case (with the `Grow` component), and this is what I see...
![image](https://user-images.githubusercontent.com/38944003/55718307-533de100-59f3-11e9-9c50-a3bf8a76389e.png)
Hopefully it's (relatively) self-explanatory

I can see that the `Portal` is indeed calling `this.forceUpdate` and re-rendering correctly, so I'm not sure why the `wrapper.find('span').length` is equal to 0, and not 1. Is it that the `wrapper` instance doesn't get sync'd with the underlying DOM unless `wrapper.update` is called (I thought that just ran a `forceUpdate` on the component tree...?)

Reply:
When is the assertion happening? Try switching to to `console.log`. 

Might be that the component tree isn't updated for async callbacks. Try `wrapper.update()` if the timing is off.

Reply:
So it looks like `wrapper.update()` only updates the component when required, and doesn't force a re-render [as it seems to say](https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L263)
(see [update doesn't force a render when none is needed. Try .forceUpdate()](https://github.com/airbnb/enzyme/issues/1833#issuecomment-423790980). I can't find a better source.)

This also seems to be validated by my very crude console.log testing...
![image](https://user-images.githubusercontent.com/38944003/55735227-d7568f80-5a18-11e9-9d61-9ed50a5dd8d1.png)
I have it logging out on every `Popper`, `Portal`, `Grow`, and `Grow.props.children` render, and those logs are only shown on the initial `setProps({ open: true })` call.

The `wrapper.find('span').length === 0` until I call `wrapper.update()`, but that is **NOT** causing a re-render - Looks like it is simply sync'ing the enzyme `ReactWrapper` with the underlying DOM (perhaps it is out-of-sync because of Portal's `forceRender`? ¯\\_(ツ)_/¯)

Ergo: This test needs to be changed to run `wrapper.update()` after the `setProps({ open: true })`, and assert that the component is mounted, but not yet got a transition state of "ENTERED"

I didn't realise this could be so confusing :/